source/_ext/gitstamp.py

```diff
@@ -0,0 +1,139 @@
+    fullpagename = pagename
+    docsrc = ''
+    try:
+        docsrc = app.confdir + "/"
+        if docsrc != "/":
+            fullpagename = docsrc + pagename
+    except KeyError:
+        pass
+
+    # Don't barf on "genindex", "search", etc
+    if not os.path.isfile("%s.rst" % fullpagename):
+        return
+
+    try:
+        line = g.log('--pretty=format:%ai,%ae,%an', '-n 1', "%s.rst" % fullpagename).split(',')
+
+        if line == "":
+            # Don't datestamp generated rst's (e.g. imapd.conf.rst)
+            # Ideally want to check their source - lib/imapoptions, etc, but
+            # that involves getting the source/output pair into the extension.
+            return
+
+        updated = line[0][:10]
+        email = line[1]
+        name = line[2]
+        context['gitstamp'] = datetime.datetime.strptime(
+                updated,
+                "%Y-%m-%d"
+            ).strftime(
+                app.config.gitstamp_fmt
+            )
+
+        user = GHCachedUser(name)
+        if gh:
+            # Look in cache first to avoid spamming GitHub
+            if email in gh_user_cache:
+                user = gh_user_cache[email]
+            else:
+                # Search GitHub and retrieve first user with matching email (if any)
+                gh_users = gh.search_users(email)
+                if gh_users.totalCount:
+                    gh_user = next(iter(gh_users))
+                    user.name = gh_user.name
+                    user.login = gh_user.login
+                    user.avatar_url = gh_user.avatar_url
+                gh_user_cache[email] = user
+        context['gitauthor'] = user.name
+        context['gitlogin'] = user.login
+        context['gitavatar'] = user.avatar_url
+
+    except git.exc.GitCommandError:
+        # File doesn't exist or something else went wrong.
+        raise errors.ExtensionError("Can't fetch git history for %s.rst." %
+                                    fullpagename)
+    except ValueError:
+        # Datestamp can't be parsed.
+        app.info("%s: Can't parse datestamp () %s ) for gitstamp, output \
+            won't have last updated time." % (pagename, updated))
+
+
+# Only add the page context handler if we're generating html
+def what_build_am_i(app):
+    global g
+    if (app.builder.format != 'html'):
+        return
+
+    try:
+        import git
+    except ImportError:
+        raise errors.ExtensionError("gitpython package not installed. \
+            Required to generate html. Please run: pip install gitpython")
+
+    try:
+        global g
+        g = git.Git('.')
+    except Exception:
+        app.info(sys.exc_info()[0])
+        app.warn("gitstamp extension enabled, but no git repository found. No \
+            git datestamps will be generated.")
+    else:
+        app.connect('html-page-context', page_context_handler)
+
+    try:
+        global gh
+        global gh_user_cache
+        token = os.getenv('GITHUB_TOKEN')
+        if token:
```

Comment:
I prefer to make the GitHub link optional. If you have configured the token and have access to GitHub, then you'll have avatars and links, otherwise you'll just have the info from the commits and everything will run just fine.
I'll add a warning just in case, though.